kong/runloop/handler.lua

```diff
@@ -201,32 +227,13 @@ do
       return nil, "error attempting to acquire build_router lock: " .. err
     end
 
-    -- lock acquired but we might not need to rebuild the router (if we
-    -- were not the first request in this process to enter this code path)
-    -- check again and rebuild if necessary
-
-    version, err = cache:get("router:version", CACHE_ROUTER_OPTS, utils.uuid)
-    if err then
-      log(ngx.CRIT, "could not ensure router is up to date: ", err)
-      return nil, err
-    end
-
-    if version == router_version then
-      return router
-    end
-
-    -- router needs to be rebuilt in this worker
-    log(DEBUG, "rebuilding router")
-
-    local ok, err = build_router(singletons.db, version)
+    local ok, router, err = pcall(fetch_router)
```

Comment:
I'm not expecting runtime errors here, I think @bungle was just being defensive (I didn't have the `pcall` in my version of the patch). From a readability perspective, I think "just-in-case" pcalls tend to throw off the reader a little (e.g. implying that there are expected runtime `errors` in the code path). But given that an `error`-causing bug in this sequence would lock down the worker, I don't mind keeping it.

Reply:
About the timeout, you mean retrying to take the semaphore or just rebuilding it anyway? I assume the latter, since AFAICT the semaphore is here to avoid duplicated work (many requests rebuilding all at once) and not to protect an inconsistency, right? (i.e. nothing breaks if a couple of requests of a worker end up rebuilding the router). Is my understanding correct?

Reply:
> Instead, falling back to rebuilding the router is probably safer to avoid leaving the worker in an inconsistent state in case the leader-request failed to rebuild said router. We could avoid the return if err == "timeout"

i feel like we discussed this and decided against it, because we end up with a case where an overly-long router rebuild (longer than the timeout) eventually results in the same stampeding behavior we're trying to avoid

Reply:
ref: https://github.com/Kong/kong/pull/3782/files#r219679253

Reply:
@hishamhm In my opinion unexpected `pcall` are confusing at the very least. There should be no reason for an error to be thrown in this codepath, _except_ for the router build itself (upon invalid configuration values, which should never occur since caught by the model validation). But using a `pcall` doesn't make the proxy more resilient (we still produce a 500 here) _and_ we lose the stacktrace which is quite annoying. Not a good choice imho.

@p0pr0ck5  Yes, I recall this discussion and I already felt this way:

> if a situation ever occurs where a particular rebuild gets slowed down, we don't prevent another request to rebuild the router on its own - on Cassandra, a subsequent router rebuild would hit another peer, for instance

> The point that I am trying to convey here is that we should follow a "best effort" practice with regards to the proxy's resiliency - at all times.

In my opinion it is more important to ensure proxy consistency and resiliency than not to stampede the database. Because the former allows for the database to be scaled appropriately, but the latter is limiting by its very own behavior. That smells of wrong design.

> Again, a retry mode can be useful because it would connect to a different Cassandra peer than the first one (due to the load balancing policy), and it certainly wouldn't hurt to retry a query on postgres either, especially if some middleware is involved between Kong and the instance, as it can be the case.

I recall we originally mentioned a more complex retry logic, but the logic we are talking about here is quite simpler, and just a fallthrough.

Reply:
Runtime errors, you never know. In this case it is a semaphore protected branch where we have pcall just in case that in most scenarios should not even run.

What I was thinking is:
1. allow using stale router until a new router is build
2. build a new router in a timer

It is like Cassandra, eventually consistent. Also whenever the build router sees that version was updated during rebuilding, it should stop building immediately and instead start again. Of course having 10 workers and 2 nodes, each starting a timer and trying to build the router at about the same time produces a lot of db queries still, and if they need to restart because of router version changes, that leads to even more requests. But I'd say it is still better than queuing the requests and occasionally not serving requests at all because router could not be built. Then we always have consistent router, only that it can be a bit stale, but it is always like that.

Reply:
> Runtime errors, you never know. 

There should be no runtime error possible on this codepath (I never recall reports of seeing one), and regardless, both the `pcall` and the runtime error produce the same response for a client, so what is the point? Besides that we also lose the stacktrace.

I feel like the pagination issue is a separate one, whose appropriate fix would be the granular rebuilds of the router (something we purposefully left out from the router development a few years ago). I would like to keep this conversation about the issue at hand only for now, and solve it appropriately. Besides, I am not even sure the faulty condition is realistic, given the cache invalidation queuing of events + grace period mechanism, a subsequent router rebuild should eventually be triggered. I recall this issue crossed my mind while developing the new cache mechanism (Since other cached items are granular, but the router isn't, it gave me quite a few challenges like this. But since that is only one of the many issues of a non-granular router, I decided to assume that an eventual rebuild should pretty much always be triggered, which was "good enough" for shipping back then). 

> occasionally not serving requests at all because router could not be built

If we fallback upon semaphore timeout, those requests are free to rebuild the router on their own.

Let's take this step by step. I like the timer idea (assuming with some semaphore logic as well maybe), but I feel like this PR is a good step forward already, and adding the fallthrough makes it even better.

Reply:
All valid points everyone, and many of the points raised match the discussion I had with @p0pr0ck5 yesterday. It is a very valid discussion, but I think its scope extends beyond the semaphore bugfix that is the goal of this PR.

I pushed an update to the PR's commit including the regression test and removing the `pcall` in the spirit of focusing on the bugfix and reducing behavioral changes here that deserve further discussion elsewhere.

I kept the unrelated `each()` change in the PR as it seems an uncontroversial improvement.

Reply:
> Besides that we also lose the stacktrace.

and also release a semaphore.

> a subsequent router rebuild should eventually be triggered

Yes, I think it is faulty only for some very unlucky leader requests.

@hishamhm I think you can remove `pcall` (I do not feel strongly against removal, just that I don't want our hot path to possibly halt processing requests for 60 secs and queuing a lot on highly traffic server — can there be an runtime error that is not in our hands, e.g. cosockets or something thowing that, there is a lot of components involved when we do rebuilding) and add `fallback`?

Reply:
> and also release a semaphore.

That is actually a very good point. So good that I think it does outweight the cons and is worth keeping it even if the likelihood of a runtime error ever happening in that code path seems very low. I am changing my mind, and would prefer that we actually keep it :+1: Sorry for the noise.

And adding the fallback :)

Reply:
Yeah, that's what I meant by "an `error`-causing bug in this sequence would lock down the worker". I'm confident about the code path as well, but if it's good for semaphore-safety and safer and most people are in for it, then yeah, let's add it. At first I was under the mistaken impression that this semaphore logic had been out for longer, but since it's rather new, the extra caution is warranted.